src/extensions/LensExtension.js

```diff
@@ -0,0 +1,130 @@
+        ...viewState,
+        viewState
+      });
+      // If the mouse is in the viewport and the mousePosition exists, set
+      // the state with the bounding box of the circle that will render as a lens.
+      if (mousePosition && viewport.containsPixel(mousePosition)) {
+        const offsetMousePosition = {
+          x: mousePosition.x - viewport.x,
+          y: mousePosition.y - viewport.y
+        };
+        const mousePositionBounds = [
+          // left
+          [offsetMousePosition.x - lensRadius, offsetMousePosition.y],
+          // bottom
+          [offsetMousePosition.x, offsetMousePosition.y + lensRadius],
+          // right
+          [offsetMousePosition.x + lensRadius, offsetMousePosition.y],
+          // top
+          [offsetMousePosition.x, offsetMousePosition.y - lensRadius]
+        ];
+        // Unproject from screen to world coordinates.
+        const unprojectLensBounds = mousePositionBounds.map(
+          (bounds, i) => viewport.unproject(bounds)[i % 2]
+        );
+        layer.setState({ unprojectLensBounds });
+      } else {
+        layer.setState({ unprojectLensBounds: [0, 0, 0, 0] });
+      }
+    };
+    if (this.context.deck) {
+      this.context.deck.eventManager.on({
+        pointermove: onMouseMove,
+        pointerleave: onMouseMove,
+        wheel: onMouseMove
+      });
+    }
+    this.setState({ onMouseMove, unprojectLensBounds: [0, 0, 0, 0] });
+  }
+
+  draw(params, extension) {
+    const { bounds } = this.props;
+    const { unprojectLensBounds } = this.state;
+    const {
+      isLensOn,
+      lensSelection,
+      lensBorderColor = [255, 255, 255],
+      lensBorderRadius = 0.02
+    } = extension.opts;
+    // Creating a unit-square scaled intersection box for rendering the lens.
+    // It is ok if these coordinates are outside the unit square since
+    // we check membership in or out of the lens on the fragment shader.
+    const [
+      leftMouseBound,
+      bottomMouseBound,
+      rightMouseBound,
+      topMouseBound
+    ] = unprojectLensBounds;
+    const [left, bottom, right, top] = bounds;
+    const leftMouseBoundScaled = (leftMouseBound - left) / (right - left);
+    const bottomMouseBoundScaled = (bottomMouseBound - top) / (bottom - top);
+    const rightMouseBoundScaled = (rightMouseBound - left) / (right - left);
+    const topMouseBoundScaled = (topMouseBound - top) / (bottom - top);
+    const uniforms = {
+      majorLensAxis: (rightMouseBoundScaled - leftMouseBoundScaled) / 2,
+      minorLensAxis: (bottomMouseBoundScaled - topMouseBoundScaled) / 2,
+      lensCenter: [
+        (rightMouseBoundScaled + leftMouseBoundScaled) / 2,
+        (bottomMouseBoundScaled + topMouseBoundScaled) / 2
+      ],
+      isLensOn,
+      lensSelection,
+      lensBorderColor,
+      lensBorderRadius
+    };
+    // ScatterplotLayer model
+    // eslint-disable-next-line no-unused-expressions
+    this.state.model?.setUniforms(uniforms);
```

Comment:
As I already commented before, `this.state.model` is `undefined` when adding this extension to a composite layer such as `MultiscaleImageLayer`. The extension would need to be added to `XRLayer` for `setUniforms` to be called. Also, at least the `ScatterplotLayer model` comment is left over from copy-pasta.

Reply:
Even if `this.state.model` is `undefined`, wouldn't the `?` operator mean this operation doesn't fail?  I'll remove the comment.

Reply:
Yes, this won't cause an error. My concern is about getting the extension to the primitive layer (here: `XRLayer` or `XR3DLayer`) where `setUniforms` would actually be called. As I just found out, an extension can implement `getSubLayerProps`. That may be a way for the extension to pass itself down to a primitive layer.

Reply:
Could you suggest a code-diff or something?  I still don't follow.  The extension gets passed into `MultiscaleImageLayer` `props` which get passed down to `XRLayer` without issue.  This doesn't sound like what you're question is.

Reply:
This turned out to be my lack of understanding of what gets passed to sub-layers. `extensions` is among the props that get passed on:
```js
  getSubLayerProps(sublayerProps = {}) {
    const {
      opacity,
      pickable,
      visible,
      parameters,
      getPolygonOffset,
      highlightedObjectIndex,
      autoHighlight,
      highlightColor,
      coordinateSystem,
      coordinateOrigin,
      wrapLongitude,
      positionFormat,
      modelMatrix,
      extensions,
      fetch,
      _subLayerProps: overridingProps
    } = this.props;
```